### PR TITLE
test: Add some selenium tests for machines

### PR DESCRIPTION
**Pull request by hahahub**

### CODE REVIEW

File: bots/machine/machine_core/ssh_connection.py [hunk 1]
```diff
@@ -233,7 +233,7 @@ def _ensure_ssh_master(self):
             self._start_ssh_master()
 
     def execute(self, command=None, script=None, input=None, environment={},
-                stdout=None, quiet=False, direct=False, timeout=120,
+                stdout=None, quiet=False, direct=False, timeout=120,raising=True,
```
Comment (martinpitt) [inline bots/machine/machine_core/ssh_connection.py]: I don't see this being used anywhere? Please remove it again, it's much more explicit to either call a command with `|| true`, or even better, an explicit error handler.

Comment (hahahub) [reply bots/machine/machine_core/ssh_connection.py]: OK, Many thanks

File: test/avocado/selenium-machines-vcpu.py [hunk 2]
```diff
@@ -75,3 +77,18 @@ def testVcpuConfig0(self):
...
+        self.click(self.wait_css('#vm-staticvm-usage', cond=clickable))
+        self.assertEqual(self.wait_css('#chart-donut-1 .donut-title-big-pf', cond=visible).text, '0.0')
+
+    @skipIf('can not get the cpu usage')
+    def testVcpuUsageRunning(self):
+        self.create_vm()
+
+        self.click(self.wait_css('#vm-staticvm-usage', cond=clickable))
+        # # wait for appearing the data
+        sleep(15)
```
Comment (KKoukiou) [inline test/avocado/selenium-machines-vcpu.py]: `sleep` with hardcoded numbers is not good practice. Please keep in mind that we run the selenium tests gating each PR, thus adding this 15 sec extra time to the tests in not desirable.

File: test/avocado/testlib_avocado/machineslib.py [hunk 3]
```diff
@@ -166,3 +167,87 @@ def tearDown(self):
         super().tearDown()
         if self.virshvm:
             self.destroy_vm(self.virshvm)
+        if len(self.vmlist) != 0:
+            for vm in self.vmlist:
+                self.destroy_vm(vm)
+        if self.machine.execute("sudo virsh pool-list | grep apool", raising=False) != '' and self.machine.execute("virsh pool-list | grep mypool", raising=False) != '':
+            self.machine.execute(
+                'sudo virsh vol-delete apooldisk1 apool \
```
Comment (KKoukiou) [inline test/avocado/testlib_avocado/machineslib.py]: If you don't want to use vol-delete for some specific reason, you can just destroy the pool (aka deactivate it) and them do rm -rf the pool directory, omiting all vol-delete commands.

File: test/avocado/testlib_avocado/machineslib.py [hunk 4]
```diff
@@ -166,3 +167,87 @@ def tearDown(self):
...
+        if self.machine.execute("sudo virsh vol-list default | grep detachdisk", raising=False) != '':
+            self.machine.execute('sudo virsh vol-delete detachdisk default')
+        if self.machine.execute('sudo cat /etc/passwd | grep auto', raising=False) != "":
+            self.machine.execute('sudo userdel -rf auto')
+
+    def prepare_disk(self):
+        self.machine.execute(
+            'sudo mkdir /home/{apool} \
+                && sudo virsh pool-create-as {apool} --type dir --target /home/{apool} \
+                && sudo virsh vol-create-as {apool} {apool}disk1 --capacity 1G --format qcow2 \
```
Comment (KKoukiou) [inline test/avocado/testlib_avocado/machineslib.py]: maybe some underscore between the pool name and disk1/disk2 would be good for readiness.
Thus having name apool_disk1 instead of apooldisk1

File: test/avocado/testlib_avocado/machineslib.py [hunk 5]
```diff
@@ -166,3 +167,87 @@ def tearDown(self):
...
+                        os=None,
+                        mem=None,
+                        mem_unit='G',
+                        storage=None,
+                        storage_unit='G',
+                        imme=False):
+        self.vmlist.append(name)
+        self.click(self.wait_css('#create-new-vm', cond=clickable))
+
+        if connection_type:
```
Comment (KKoukiou) [inline test/avocado/testlib_avocado/machineslib.py]: These two 'pass' here are redundant.

File: test/avocado/testlib_avocado/machineslib.py [hunk 6]
```diff
@@ -166,3 +167,87 @@ def tearDown(self):
...
+            self.click(self.wait_css('#memory-size-unit-select li:nth-child(1) a'))
+        if storage:
+            self.send_keys(self.wait_css('#storage-size'), storage)
+        if storage_unit != 'G':
+            self.click(self.wait_css('#storage-size-unit-select button', cond=clickable))
+            self.click(self.wait_css('#storage-size-unit-select li:nth-child(1) a', cond=clickable))
+        if imme:
+            self.click(self.wait_css('#start-vm', cond=clickable))
+        # for waiting the element has right input
+        sleep(2)
```
Comment (KKoukiou) [inline test/avocado/testlib_avocado/machineslib.py]: Please no hardcoded sleeps.

Comment (KKoukiou) [reply test/avocado/testlib_avocado/machineslib.py]: There are many function that allow you to wait on some html/css for appear. wait_present, wait_visilbe, wait_attr please use these instead of sleeps.

File: test/avocado/testlib_avocado/machineslib.py [hunk 7]
```diff
@@ -166,3 +167,87 @@ def tearDown(self):
...
+        self.click(self.wait_css('#create-vm-dialog .btn-primary', cond=clickable))
+        self.wait_dialog_disappear()
+
+    def wait_dialog_disappear(self):
+        while self.wait_css('#app').get_attribute('aria-hidden'):
```
Comment (KKoukiou) [inline test/avocado/testlib_avocado/machineslib.py]: Is this really the proper way to find out that a dialog went away?
How I was doing it so far is that I was checking for the '.modal' class to not be present either I was checking the modal id, where is was available.

File: test/avocado/selenium-machines-basic.py [hunk 8]
```diff
@@ -107,3 +107,71 @@ def testDelete(self):
...
+        self.wait_css('#vm-staticvm-disks-vda-bus')
+
+        self.click(self.wait_css('#vm-staticvm-delete', cond=clickable))
+        self.click(self.wait_css("#vm-staticvm-delete-modal-dialog button.btn-danger", cond=clickable))
+
+        self.wait_dialog_disappear()
+        self.wait_css('#vm-staticvm-row', cond=invisible)
+        self.assertNotIn('staticvm', self.machine.execute('sudo virsh list --all'))
+        self.assertNotIn(args.get('image',None), self.machine.execute('sudo virsh vol-list default'))
+        self.assertNotIn(attach_img, self.machine.execute('sudo virsh vol-list default'))
```
Comment (KKoukiou) [inline test/avocado/selenium-machines-basic.py]: What about the back_img? Don't you want to check for it being deleted?
Also please don't rely on the checkboxes being set, you have to check the checkboxes status as well.

**Comment by hahahub** [comment]
OK, i will recheck and modify the code.Sorry for my mistake.

**Comment by martinpitt** [comment]
@hahahub : Please just force-push to your branch, let's keep this PR open. Otherwise the discussion gets spread out over multiple PRs and hard to follow.

**Comment by hahahub** [comment]
@martinpitt, i'm afraid i need to close the PR. Because this is my first PR, and i didn't realize that i need my own branch.So, i think it's better to repost the PR after some time, then i will recheck and modify the code, and make them in two commits(seleniumlib and new tests).

**Comment by martinpitt** [comment]
OK, as you wish. Thanks!
